drivers/ninaw10/nina_wifi_drv.c

```diff
@@ -752,23 +753,37 @@ int nina_socket_listen(int fd, uint32_t backlog) {
     return 0; // No listen ?
 }
 
-int nina_socket_accept(int fd, uint8_t *ip, uint16_t *port, int *fd_out, uint32_t timeout) {
+int nina_socket_avail(int fd, int type, uint16_t *data) {
     uint16_t size = 2;
-    uint16_t sock = NO_SOCKET_AVAIL;
+
+    if (nina_send_command_read_vals(NINA_CMD_SOCKET_AVAIL,
+        1, ARG_8BITS, NINA_ARGS(ARG_BYTE(fd)),
+        1, ARG_8BITS, NINA_VALS({&size, data})) != 0) {
+        return -1;
+    }
+
+    // For TCP sockets in listen state, return 0 if there's no accepted socket.
+    if (*data == NO_SOCKET_AVAIL && type == NINA_SOCKET_TYPE_TCP
+        && nina_server_socket_status(fd) == SOCKET_STATE_LISTEN) {
+        *data = 0;
+    }
+
+    return 0;
+}
+
+int nina_socket_accept(int fd, uint8_t *ip, uint16_t *port, int *fd_out, int32_t timeout) {
+    uint16_t sock = 0;
 
     if (nina_server_socket_status(fd) != SOCKET_STATE_LISTEN) {
         return -1;
     }
 
-    for (mp_uint_t start = mp_hal_ticks_ms(); sock == 0 || sock == NO_SOCKET_AVAIL; mp_hal_delay_ms(10)) {
-        if (nina_send_command_read_vals(NINA_CMD_SOCKET_ACCEPT,
-            1, ARG_8BITS, NINA_ARGS(ARG_BYTE(fd)),
-            1, ARG_8BITS, NINA_VALS({&size, &sock})) != 0) {
+    for (mp_uint_t start = mp_hal_ticks_ms(); !sock; mp_hal_delay_ms(10)) {
+        if (nina_socket_avail(fd, NINA_SOCKET_TYPE_TCP, &sock) != 0) {
             return -1;
         }
-
-        if (timeout && (mp_hal_ticks_ms() - start) >= timeout) {
-            return NINA_ERROR_TIMEOUT;
+        if (timeout == 0 || (timeout > 0 && (mp_hal_ticks_ms() - start) >= timeout)) {
```

Comment:
It looks like this returns -MP_ETIMEDOUT even if there is data available: if the above call to `nina_socket_avail` returns non-zero in `sock` (which should be called `avail`) then that's not checked until the next for loop iteration, and if `timeout==0` that will never occur.

Reply:
And below in `NINA_CMD_SOCKET_REMOTE_ADDR`, is `sock` supposed to be passed as the first argument?  Below it seems to be used as the new fd...

Reply:
Fixed.

Reply:
> nina_socket_avail returns non-zero in sock (which should be called avail)

What should be called `avail` ?

> And below in NINA_CMD_SOCKET_REMOTE_ADDR, is sock supposed to be passed as the first argument? Below it seems to be used as the new fd...

Yes, it returns the remote IP/PORT, and it is the accepted socket.

Reply:
But what about `sock` below, it's now the number of bytes available on the socket, when it should be the new file descriptor of the newly accepted socket.

Reply:
`nina_socket_avail()` or this command `NINA_CMD_SOCKET_AVAIL`, returns the number of bytes available or a newly accepted socket if the parent socket is in listen state.

Reply:
Then, `sock` is used in `NINA_CMD_SOCKET_REMOTE_ADDR` to return the accepted socket's IP and port.

Reply:
It's confusing, but it's the way the protocol is designed, it could've been much better for sure. It took me a while to figure out some of the vague commands inner workings by looking at the firmware source.

Reply:
ok!